### PR TITLE
Rewrite dependency injection to use Dagger

### PR DESCRIPTION
**Pull request by msfjarvis**

### CODE REVIEW

File: app/src/main/java/com/wireguard/android/model/TunnelManager.kt [hunk 1]
```diff
@@ -28,10 +28,11 @@ import com.wireguard.config.Config
 import java9.util.Comparators
 import java9.util.concurrent.CompletableFuture
 import java9.util.concurrent.CompletionStage
+import javax.inject.Inject
 import org.koin.core.KoinComponent
 import timber.log.Timber
 
-class TunnelManager(
+class TunnelManager @Inject constructor(
```
Comment (msasikanth) [inline app/src/main/java/com/wireguard/android/model/TunnelManager.kt]: You can make the class `@Reusable`

File: app/src/main/java/com/wireguard/android/util/AsyncWorker.kt [hunk 2]
```diff
@@ -9,12 +9,13 @@ import android.os.Handler
...
 import java9.util.concurrent.CompletableFuture
 import java9.util.concurrent.CompletionStage
+import javax.inject.Inject
 
 /**
  * Helper class for running asynchronous tasks and ensuring they are completed on the main thread.
  */
 
-class AsyncWorker(private val executor: Executor, private val handler: Handler) {
+class AsyncWorker @Inject constructor(private val executor: Executor, private val handler: Handler) {
```
Comment (msasikanth) [inline app/src/main/java/com/wireguard/android/util/AsyncWorker.kt]: You can make the class `@Reusable`

File: app/src/main/java/com/wireguard/android/util/RootShell.kt [hunk 3]
```diff
@@ -16,9 +16,10 @@ import java.io.InputStreamReader
 import java.io.OutputStreamWriter
 import java.nio.charset.StandardCharsets
 import java.util.UUID
+import javax.inject.Inject
 import timber.log.Timber
 
-class RootShell(val context: Context) {
+class RootShell @Inject constructor(val context: Context) {
```
Comment (msasikanth) [inline app/src/main/java/com/wireguard/android/util/RootShell.kt]: You can make the class `@Reusable`

File: app/src/main/java/com/wireguard/android/util/ToolsInstaller.kt [hunk 4]
```diff
@@ -12,14 +12,15 @@ import com.wireguard.android.util.RootShell.NoRootException
...
+import javax.inject.Inject
 import org.koin.core.KoinComponent
 import timber.log.Timber
 
 /**
  * Helper to install WireGuard tools to the system partition.
  */
 
-class ToolsInstaller(private val context: Context, private val rootShell: RootShell) : KoinComponent {
+class ToolsInstaller @Inject constructor(private val context: Context, private val rootShell: RootShell) : KoinComponent {
```
Comment (msasikanth) [inline app/src/main/java/com/wireguard/android/util/ToolsInstaller.kt]: You can make the class `@Reusable`

**Comment by msfjarvis** [comment]
Takeaways from this effort:

- Dagger is ridiculously complex without a human to guide you around.
- Koin's service locator pattern makes it far too easy to write bad code because you can inject anything anywhere.
- Dagger can feel like a lot of boilerplate but some clever techniques can mitigate that.
- Writing `inject` methods for every single class can feel like a drag because it is.
- Injecting into Kotlin `object`s appears to be a no-go. I opted to [refactor out the staticity where possible](https://github.com/msfjarvis/viscerion/pull/214/commits/9eb532521f51d0f7bb66a2a78aa1fc5688128a22), [pass injected dependencies to the function](https://github.com/msfjarvis/viscerion/commit/e23f878140d4bda9e2c54d6c2684e07994066fd6#diff-28007a5799b03e7b556f5bb942754031) or [fall back to 'dirty' patterns](https://github.com/msfjarvis/viscerion/pull/214/commits/fc54ec6bb8e99ec639c6617765e814e12d91ea1a#diff-74f75ab44e1cd2909c4ec4d704bbbab7R65) as needed. Do what you feel like.
- I still do not _love_ Dagger. Fuck you Google.

**Pull request merged**